Source/NSData.m

```diff
@@ -1082,6 +1082,119 @@ - (NSData*) subdataWithRange: (NSRange)aRange
   return [NSData dataWithBytesNoCopy: buffer length: aRange.length];
 }
 
+/**
+ * Finds and returns the range of the first occurrence of the given data, within the given range, subject to given options.
+ */
+- (NSRange) rangeOfData: (NSData *)dataToFind
+                options: (NSDataSearchOptions)mask
+                  range: (NSRange)searchRange
+{
+  NSUInteger  length = [self length];
+  NSUInteger  countOther = [dataToFind length];
+  const void* bytesSelf = [self bytes];
+  const void* bytesOther = [dataToFind bytes];
+  NSRange     result;
+
+  GS_RANGE_CHECK(searchRange, length);
+  if (dataToFind == nil)
+    [NSException raise: NSInvalidArgumentException format: @"range of nil"];
+
+  /* Zero length data is always found at the start of the given range.
+   */
+  if (0 == countOther)
+    {
+      if ((mask & NSDataSearchBackwards) == NSDataSearchBackwards)
+        {
+          searchRange.location += searchRange.length;
+        }
+      searchRange.length = 0;
+      return searchRange;
+    }
+
+  if (searchRange.length < countOther)
+    {
+      /* Range to search is smaller than data to look for.
+       */
+      result = NSMakeRange(NSNotFound, 0);
+    }
+  else
+    {
+      if ((mask & NSDataSearchAnchored) == NSDataSearchAnchored
+        || searchRange.length == countOther)
```

Comment:
Or do you mean `&&` here?

Reply:
Similar to the prior comment, this was adapted directly from rangeOfString, and I think the code is assuming that an anchored search is the same as a search where self and other are the same length (of course just not caring if it's not true for self).  It does seem off with regard to a backwards search, so I'll probably have better feedback here after addressing the prior comment.